node/impl/full/eth.go

```diff
@@ -1703,6 +1762,68 @@ func newEthTxReceipt(ctx context.Context, tx ethtypes.EthTx, lookup *api.MsgLook
 	return receipt, nil
 }
 
+func (m EthTxHashManager) Apply(ctx context.Context, from, to *types.TipSet) error {
+	for _, blk := range to.Blocks() {
+		_, smsgs, err := m.StateAPI.Chain.MessagesForBlock(ctx, blk)
+		if err != nil {
+			return err
+		}
+
+		for _, smsg := range smsgs {
+			if smsg.Signature.Type != crypto.SigTypeDelegated {
+				continue
+			}
+
+			hash, err := EthTxHashFromSignedFilecoinMessage(ctx, smsg, m.StateAPI)
+			if err != nil {
+				return err
+			}
+
+			err = m.TransactionHashLookup.InsertTxHash(hash, smsg.Cid(), int64(to.Height()))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+type EthTxHashManager struct {
+	StateAPI              StateAPI
+	TransactionHashLookup *ethhashlookup.TransactionHashLookup
+}
+
+func (m EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) error {
+	return nil
+}
```

Comment:
The edge case where a reorg happens and messages get included in a future epoch is handled by the ON CONFLICT clause, which is great. However, if a message is added in tipset 1A, then we reorg to tipset 1B which does not include the message, and it only lands in tipset 4A, we would've spent 3 tipsets thinking that the message exists and it landed on epoch 1. Is that correct?

Similarly, there's the edge case where the chain reorgs and completely removes a message such that it never again lands on chain.

I'm not sure how important consistency in such scenarios is -- it might be acceptable to have wrong data in the index if the epoch is only used for garbage collection. But we should _never_ return data over the interface that is not truthful!

Reply:
I think I want to use just raw timestamps instead of epochs here. The column should just be for optionally periodically clearing out the db so users can manage the size of their state. The problem with the current setup is that we don't know when to clear out mpool mappings if they're removed. We don't need the epoch information in this table, we can look it up once we get the cid. 

If we're not storing the epoch, I don't think there are any inconsistency issues.

Reply:
AFAIK once a transaction is successfully added in the mpool (has passed validation), I believe that four logical outcomes are possible:
1. It can land on chain, and is eventually finalized.
2. It can land on chain and then be reverted due to a reorg, "returning" to the mpool.
3. It can be replaced.
4. It can drop.

Are we handling all of these? Wanna make sure that there's no possible issue here with stale or bad data in the cache?

Reply:
We don't explicitly remove mappings for messages in scenarios 2-4, so we might end up with a few mappings in the database for transactions that never fully land on chain. I don't think this is an issue. I've removed the epoch which could've led to inconsistencies. There's no harm in returning a cid for a message that is no longer relevant - the caller will look up the cid and will be unable to find it in their store.

Reply:
@geoff-vball are you saying that we have covered the proper use cases? To your point, stale or useless messages are likely harmless to leave around outside of the space it takes up. When it comes to getting a canonical version of the data, I imagine that comes from a snapshot?

Either way, being able to garbage collect that information seems orthogonal to the criticality of how re-orgs are handled.

Are you saying that we are safe against the 4 considered cases? Is there any other blocker?

Reply:
Yeah, there are no situations where you'd be missing something from your DB, unless it has been garbage collected by the retention policy, or you had the DB turned off when you synced.

Reply:
Just to summarize how it works: 
Adds mapping when:
- Tx added to mpool
- Tx processed in block

Removes mappings when:
- Checks every hour for mappings that have been in the database longer than the retention policy.

Reply:
What is the default retention policy? Curious.

Reply:
@scotthconner Default sets the database to not garbage collect at all. If they notice their db getting too large they can change the retention policy. 

Thinking about this some more, I think the database should mirror the retention policy of messages in the blockstore. The mappings become (relatively) useless if we prune a message from our store. You'll be able to find the cid, but you won't be able to use to cid to look up the message. 

You could theoretically use that cid to look up the transaction on filfox or filscan. Do we know if they're planning on implementing lookups by Transaction Hashes?